src/popover/container.tsx

```diff
@@ -47,145 +48,186 @@ export default function PopoverContainer({
+      const trackRelativeOffset = toRelativePosition(popoverOffset, toRelativePosition(trackRect, containingBlockRect));
+
+      // Bring back the container to its original position to prevent any flashing.
+      popover.style.top = prevTop;
+      popover.style.left = prevLeft;
+
+      // Allow popover body to scroll if can't fit the popover into the container/viewport otherwise.
+      if (scrollable) {
+        body.style.maxHeight = Math.min(boundingOffset.height, window.innerHeight - parseInt(prevTop)) + 'px';
+        body.style.overflowX = 'hidden';
+        body.style.overflowY = 'auto';
+      }
+
+      // Position the popover
+      internalPositionRef.current = newInternalPosition;
+      setPopoverStyle({ top: popoverOffset.top, left: popoverOffset.left });
+
+      positionHandlerRef.current = () => {
+        const newTrackOffset = toRelativePosition(
+          track.getBoundingClientRect(),
+          containingBlock ? containingBlock.getBoundingClientRect() : viewportRect
+        );
+        setPopoverStyle({
+          top: newTrackOffset.top + trackRelativeOffset.top,
+          left: newTrackOffset.left + trackRelativeOffset.left,
+        });
+      };
+    },
+    [trackRef, keepPositionOnResize, position, renderWithPortal]
+  );
 
   // Recalculate position when properties change.
   useLayoutEffect(() => {
     updatePositionHandler();
   }, [updatePositionHandler, trackKey]);
 
   // Recalculate position when content size changes.
-  useResizeObserver(contentRef, () => updatePositionHandler());
+  useResizeObserver(contentRef, () => {
+    updatePositionHandler(true);
+  });
 
   // Recalculate position on DOM events.
   useLayoutEffect(() => {
     /*
-      This is a heuristic. Some layout changes are caused by user clicks (e.g. toggling the tools panel, submitting a form),
-      and by tracking the click event we can adapt the popover's position to the new layout.
-
-      TODO: extend this to Enter and Spacebar?
+    This is a heuristic. Some layout changes are caused by user clicks (e.g. toggling the tools panel, submitting a form),
+    and by tracking the click event we can adapt the popover's position to the new layout.
     */
-    const updatePosition = () => requestAnimationFrame(() => updatePositionHandler());
+
+    const onClick = (event: UIEvent | KeyboardEvent) => {
+      if (
+        // Do not update position if the click was on the popover itself
+        // and keepPositionOnResize is true.
+        (keepPositionOnResize && nodeContains(popoverRef.current, event.target)) ||
+        // If the click was on the trigger, this will make the popover appear or disappear,
+        // so no need to update its position either in this case.
+        nodeContains(trackRef.current, event.target)
+      ) {
+        return;
+      }
+
+      requestAnimationFrame(() => {
+        updatePositionHandler();
+      });
+    };
+
+    const onKeyUp = (event: KeyboardEvent) => {
```

Comment:
If the click event is no longer recorded, does it make sense to introduce the new onKeyUp handler in scope of this PR?

Reply:
Just removed. In fact, not every space or enter event is equivalent to a click (the user could be typing in an input, for example).